### PR TITLE
Gui: Use FreeCAD Theme only for FreeCAD stuff

### PR DESCRIPTION
**Pull request by kadet1090**
This should fix the issue with FreeCAD icons leaking to system dialogs. Icons loaded with BitmapFactory (almost all) will respect the new user parameter `UseIconTheme` under `BaseApp/Preferences/Bitmaps/Theme` group. Setting this parameter to true will force Qt to use icons from icon theme - by default the system one. Icon theme can be overriden via `Name` parameter. The default is false so FreeCAD will use the default icons. It does not touch Qt icon theme mechanisms so system dialogs, buttons etc should still use the icons from system theme.

Fixes: https://github.com/FreeCAD/FreeCAD/issues/16020

### CODE REVIEW

**Comment by wwmayer** [comment]
With the recent changes on themes the whole mechanism has become way too complicated and this PR makes things even more complicated and convoluted.

The whole problems (#15540, #16020) started with this PR https://github.com/FreeCAD/FreeCAD/pull/14970 where effectively system icons are replaced by FreeCAD icons.

When looking at #14968 I can understand that in your case the system icons don't fit very well. But a much less problematic way would have been to change the Name parameter (under BaseApp/Preferences/Bitmaps/Theme) to `FreeCAD-default` because then the FreeCAD icons are used without affecting the system themes.

Btw, for me the system icons fit very well to the rest of the FreeCAD icons and since I don't like the new New, Open, Save, Undo, Redo, ... icons I didn't want to have FreeCAD-default as the default theme on Linux.

I can accept FreeCAD-default as the default theme but then there must be an easier way without weird hacks to switch to the system theme.

**Comment by kadet1090** [comment]
Setting the name to FreeCAD default stopped working for some recent versions of the Qt framemework. I found only two possible solutions: 
1. Creating the FreeCAD-default theme according to the xdg standard (which was done in #14970) which works also for windows 11 (mentioned by Chris) but results in using icons everywhere including system dialogs.
2. Supplying some random theme name to force usage of fallbacks which is ugly hack and does not work for windows.

Screenshot Included in my issue is not from my setup - I can force my setup to work properly - it is from YouTube content meaning that more people can encounter it. People that won't tinker in settings to solve such issues. That makes it a major issue as the icons are almost invisible. 

If someone wants to have icons from their system theme - sure but it must be opt-in due to mentioned issue.

I so however agree that this adds yet another layer of tape  and I am not happy about it so.I am absolutely open to suggestions

**Comment by MisterMakerNL** [comment]
We could steal linkstage icon options.
So we have a system in place to set icons.
Which awesome.
The default icons are on Windows terrible they look to be design to look modern without being practical.

**Comment by wwmayer** [comment]
> Setting the name to FreeCAD default stopped working for some recent versions of the Qt framemework

No, this has nothing to do with Qt. I have a self-compiled FreeCAD version v0.21 that links to the same Qt version as my v0.22 build and for the v0.21 build the FreeCAD-default theme works as expected.

I have even an older v0.22 build (from commit d12635246) and there the FreeCAD-default theme works, too.

So, it's clearly a change in the FreeCAD code that causes the theme to fail.

**Comment by MisterMakerNL** [comment]
https://github.com/FreeCAD/FreeCAD/pull/15687

**Comment by kadet1090** [comment]
> No, this has nothing to do with Qt. I have a self-compiled FreeCAD version v0.21 that links to the same Qt version as my v0.22 build and for the v0.21 build the FreeCAD-default theme works as expected.

For me the `Bitmaps/Theme/ThemeSearchPaths` property ceased to work around may/june which was before https://github.com/FreeCAD/FreeCAD/pull/14970 was created and merged. I checked out revision from around march to test if it was result of some change in FreeCAD code and it did not work on any of older revisions - so clearly it is not because change in our code. Qt however got updated on my syastem around same time, @obelisk79 and few others also had the same experience IIRC. Qt 6.7 changed icon loading quite substantially looking at the code and commits (ex. https://github.com/qt/qtbase/commit/a452e2254644ffbed289fdf051eaf41d7e6a3b0d) so it might have impact.

**Comment by wwmayer** [comment]
> This PR might be important here:

As far as I can see this PR is not related.

**Comment by wwmayer** [comment]
Doing a manual bisect clearly shows that these two PRs caused the problems for me (and probably @berberic2): #14970 and #15588

After doing some more testing I can say that (at least for me) the main problems are:
1. the changes made to resource.qrc with the `<qresource prefix="/icons/FreeCAD-default/scalable">` 
2. the call of `QIcon::setFallbackThemeName(QIcon::themeName());`. If this call is disabled then the changes made to resource.qrc are not needed any more
3. the fact that there is not an easy way to force to use the system theme

This was tested on Ubuntu 22.04 with Qt 5.15.3 and Qt 6.2.4

To solve the issues the following needs to be done:
* The same effect as in 1. can be achieved if the path `:/icons` is added to the theme search path as it was in the past -- in case this is needed on some systems (not needed here).
* There must be a boolean user parameter to control the call of `QIcon::setFallbackThemeName(QIcon::themeName());`
* #15990 implements an easier way to force the use of the system theme

**Comment by kadet1090** [comment]
The PRs that you are mentioning were merged after revision I am mentioning https://github.com/FreeCAD/FreeCAD/pull/16018#issuecomment-2308392700 here. The old way of forcing use of FreeCAD provided icons stopped working with release of Qt 6.7.x which is the Qt version for arch based distros. You are testing on Qt version which does not implement changes I mentioned and thus it is working fine for you.

To fix this for Qt versions newer than 6.7 I found only two working ways mentioned in https://github.com/FreeCAD/FreeCAD/pull/16018#issuecomment-2307276077. Chosen way also fixes icons on windows 11 mentioned by Chris in https://github.com/FreeCAD/FreeCAD/pull/14970#issuecomment-2192773307 which earlier was not a problem as windows did not support icon themes anyway - but it does from Qt 6.7: https://github.com/qt/qtbase/commit/d2e163d2e4c388a42a006ddd43c58bbecc3729a9 

 >There must be a boolean user parameter to control the call of QIcon::setFallbackThemeName(QIcon::themeName());
With this PR this line is no longer needed.

>https://github.com/FreeCAD/FreeCAD/pull/15990 implements an easier way to force the use of the system theme
This PR still allows either using system icons whenever possible (which results in https://github.com/FreeCAD/FreeCAD/issues/14968) or to use FreeCAD icons everywhere (which results in https://github.com/FreeCAD/FreeCAD/issues/16020).

With this PR you simply change the `UseIconTheme` to false and get the default i.e. system icons. This is also quite easy. Additionally by default you get the FreeCAD icons in toolbars etc and for buttons / system dialogs (stuff controlled by the Qt) you still get icons from system theme. Yes, it comes at a price of a bit more complexity in the BitmapFactory but it is the only way to fix both #14968 and #16020 at the same time for all Qt versions.

**Comment by obelisk79** [comment]
The @FreeCAD/design-working-group was discussing this issue the other day on Discord and I was asked by a few others to share my thoughts here. 

Using system icons for just a few buttons in FreeCAD when FreeCAD has a unique art style means that there will be many types of visual inconsistency across platforms, systems, system themes etc. The decision to ever allow/use the system icons in FreeCAD's toolbars doesn't make much sense in the scope of application use. While it may work well in a few specific circumstances for some people, many others are left with a very poor experience. We should be considering a holistic vision for the interface, and start moving away from, what at the user facing level, feels like slapstick design choices. This is the same concept as wanting to avoid slapstick quality code contributions.

**Comment by yorikvanhavre** [comment]
The thing is, on Linux, there are many really nice and complete Qt themes available, and it's very much part of a great Linux desktop experience to have everything integrated and the same system icons used everywhere.

Basically me and @wwmayer (and I guess many other linux users) would like to keep the option to use the system theme defaults. We are both OK with changing the default, but it's also my opinion that it should still be available as a (reasonably easy) option.

**Comment by wwmayer** [comment]
> The thing is, on Linux, there are many really nice and complete Qt themes available, and it's very much part of a great Linux desktop experience to have everything integrated and the same system icons used everywhere.

+1

**Comment by MisterMakerNL** [comment]
QT pretty much killed off Classic for windows.
Since it is broken on windows 11 and the solution is to us QT style Fusion.
Applying QT style fusion would break the Linux integration and the macOS integration.
Using our own stylesheets would fix all this.

**Comment by kadet1090** [comment]
> The thing is, on Linux, there are many really nice and complete Qt themes available, and it's very much part of a great Linux desktop experience to have everything integrated and the same system icons used everywhere.
> 
> Basically me and @wwmayer (and I guess many other linux users) would like to keep the option to use the system theme defaults. We are both OK with changing the default, but it's also my opinion that it should still be available as a (reasonably easy) option.

This PR **does not disable or in any way make it harder** to use system icon theme, on a contrary it re-enables this ability using `BaseApp/Preferences/Bitmaps/Theme/UseIconTheme` which when set to `true` uses Qt mechanism for providing icons which by default will use the system icon theme. This is **exactly** the same step that was required to disable system provided icons in previous versions, only the parameter name was changed to better reflect use case. This should not be a problem as we agree that by default we should use icons provided by FreeCAD.

**Comment by obelisk79** [comment]
@kadet1090 I believe Yorik was kindly disagreeing my comment above about the use of system icons in an application full of custom icons being nonsensical.

**Comment by chennes** [comment]
Nonsensical or not, this option *is* available now, via the suggested UseIconTheme parameter. So I believe that resolves the issue for everyone, right? Linux users like @yorikvanhavre and @wwmayer can use their system icons by manually setting that parameter, but it's not the default so the out-of-box experience is the consistent-internal-icons that the DWG advocates for. My read of this discussion is that we have an agreeable compromise for everyone, but please weigh in if that is not the case.

**Comment by yorikvanhavre** [comment]
Since this seems to fix an urgent problem (FreeCAD icons in the KDE dialogs), it might be a good idea to merge it for 1.0, then we look for a better way to fix the system/custom them problem. Would that be ok to you @wwmayer ?

**Comment by kadet1090** [comment]
The only improvement here in regards of system icons use that I can see here to be made is to expose this to preferences dialog as a checkbox. I'm absolutely open for contributing such change but it would require to create new translation strings. 

That said, I think that it is absolutely "Fine Tuning" properties for power users and it does not need checkbox in the preferences UI as probably most users won't care much.

**Comment by chennes** [comment]
@wwmayer can we get final confirmation from you that you are OK with this solution, at least for the time being?

**Comment by wwmayer** [comment]
Go ahead.

**Comment by chennes** [comment]
@kadet1090 we can't cherry-pick onto main, can you please create another PR against main now?

**Comment by furgo16** [comment]
@kadet1090 added documentation on https://wiki.freecad.org/Fine-tuning and https://wiki.freecad.org/Release_notes_1.0 please kindly check if it's correct. 

> Icon theme can be overriden via `Name` parameter.

I did not add that to the wiki, as I wasn't sure how that parameter is supposed to work and which the full path is. Could you ellaborate? Thanks.

**Comment by MisterMakerNL** [comment]
There is already an easier way to set the icons, just add the directory with the SVG files to the customize directory.
image
image

**Comment by kadet1090** [comment]
> @kadet1090 Hey! I think this broke the IconThemes extension. I can't use the FlatIcons from @obelisk79 under windows 11 anymore. On RC01 it works, RC02 it doesn't. Can you please, look into it?

Try with `UseIconTheme` set to true, it should work

**Comment by furgo16** [comment]
After delving into theming within the snap package, I've come across this again. I added the `Name` parameter and described differences between `ThemeSearchPaths` and `UseIconTheme`.

@kadet1090 @wwmayer @Roy-043 would you mind checking if the info on the wiki for those three parameters at https://wiki.freecad.org/Fine-tuning#General is now correct? Thanks!

**Comment by kadet1090** [comment]
I'd rephrase `Name` parameter:

>Bitmaps/Theme/Name (string): Allows user to override system icon theme used by FreeCAD. Should be used only when `Bitmaps/Theme/UseIconTheme` is set to true, otherwise has no effect.

IconTheme is feature of Qt which is used only when `UseIconTheme` is enabled. The `Name` parameter allows users to override icon theme that is used by default in user system.

**Comment by Roy-043** [comment]
Thanks. I have updated the WIki page.
